airflow/security/tokens.py

```diff
@@ -0,0 +1,504 @@
+        if "kid" not in header:
+            raise ValueError("Missing 'kid' in token header")
+        return header["kid"]
+
+    async def _get_validation_key(self, unvalidated: str) -> str | jwt.PyJWK:
+        if self.secret_key:
+            return self.secret_key
+
+        if TYPE_CHECKING:
+            assert self.jwks is not None
+
+        kid = self._get_kid_from_header(unvalidated)
+        return await self.jwks.get_key(kid)
+
+    def validated_claims(
+        self, unvalidated: str, extra_claims: dict[str, Any] | None = None
+    ) -> dict[str, Any]:
+        return async_to_sync(self.avalidated_claims)(unvalidated, extra_claims)
+
+    async def avalidated_claims(
+        self, unvalidated: str, extra_claims: dict[str, Any] | None = None
+    ) -> dict[str, Any]:
+        """Decode the JWT token, returning the validated claims or raising an exception."""
+        key = await self._get_validation_key(unvalidated)
+        claims = jwt.decode(
+            unvalidated,
+            key,
+            audience=self.audience,
+            issuer=self.issuer,
+            options={"require": self.required_claims},
+            algorithms=self.algorithm,
+            leeway=self.leeway,
+        )
+
+        # Validate additional claims if provided
+        if extra_claims:
+            for claim, expected_value in extra_claims.items():
+                if expected_value["essential"] and (
+                    claim not in claims or claims[claim] != expected_value["value"]
+                ):
+                    raise InvalidClaimError(claim)
+
+        return claims
+
+
+def _pem_to_key(pem_data: str | bytes | AllowedPrivateKeys) -> AllowedPrivateKeys:
+    if isinstance(pem_data, str):
+        pem_data = pem_data.encode()
+    elif not isinstance(pem_data, bytes):
+        # Assume it's already a key object
+        return pem_data
+
+    return load_pem_private_key(pem_data, password=None)  # type: ignore[return-value]
+
+
+@attrs.define(repr=False, kw_only=True)
+class JWTGenerator:
+    """Generate JWT tokens."""
+
+    _private_key: AllowedPrivateKeys | None = attrs.field(
+        repr=False, alias="private_key", converter=_pem_to_key
+    )
+    """
+    Private key to sign generated tokens.
+
+    Should be either a private key object from the cryptography module, or a PEM-encoded byte string
+    """
+    _secret_key: str | None = attrs.field(
+        repr=False,
+        alias="secret_key",
+        default=None,
+        converter=lambda v: None if v == "" else v,
+    )
+    """A pre-shared secret key to sign tokens with symmetric encryption"""
+
+    kid: str = attrs.field()
+    valid_for: timedelta = attrs.field(converter=_sec_to_timedelta)
+    audience: str
+    issuer: str | list[str] | None = attrs.field(
+        factory=_conf_list_factory("api_auth", "jwt_issuer", first_only=True, fallback=None)
+    )
+    algorithm: str = attrs.field(factory=_conf_factory("api_auth", "jwt_algorithm", fallback="GUESS"))
```

Comment:
`algorithm` is str in `JWTGenerator` but a `list[str]` in `JWTValidator`. I think that's confusing and we should have `str` everywher, and just do the convertion when calling the decode function if needed.

Reply:
JWTValidator takes a list because if you are using other keys then you can accept multiple possible algorithms, but you can only ever generate a single algorithm type.